src/bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,363 @@
+from abc import abstractmethod
+from typing import Dict, List
+
+import numpy as np
+import pandas as pd
+from event_model import DocumentRouter, RunRouter
+from event_model.documents import EventDescriptor, StreamDatum, StreamResource
+from pydantic.utils import deep_update
+from tiled.client import from_profile, from_uri
+from tiled.client.utils import handle_error
+from tiled.structures.array import ArrayStructure, BuiltinDtype
+from tiled.structures.core import Spec, StructureFamily
+from tiled.structures.data_source import Asset, DataSource, Management
+from tiled.structures.table import TableStructure
+from tiled.utils import safe_json_dump
+
+MIMETYPE_LOOKUP = {
+    "hdf5": "application/x-hdf5",
+    "ADHDF5_SWMR_STREAM": "application/x-hdf5",
+    "AD_HDF5_SWMR_SLICE": "application/x-hdf5",
+    "AD_TIFF": "multipart/related;type=image/tiff",
+}
+DTYPE_LOOKUP = {"number": "<f8", "array": "<f8", "boolean": "bool", "string": "str", "integer": "int"}
+
+
+# TODO (probably in a future PR?):
+# Move these into external repo (probably area-detector-handlers)
+# and use the existing handler discovery mechanism.
+
+
+class StreamHandlerBase:
+    """
+    A Handler consumes documents from Bluesky. It composes details (DataSource and its Assets) that will go into
+    the Tiled database. Each Handler (more specifically, a StreamHandler) is instantiated per a Stream Resource.
+
+    Tiled Adapters will later use this to read the data, with good random access and bulk access support.
+
+    We put this code into handlers so that additional, possibly very unusual, formats can be supported by users
+    without getting a PR merged into Bluesky or Tiled.
+
+    The STREAM_HANDLER_REGISTRY (see example below) and the Tiled catalog paramter adapters_by_mimetype can be used
+    together to support:
+        - ingesting a new mimetype from Bluesky documents and generating DataSource and Asset with appropriate
+          parameters (the handler's job);
+        - Interpreting those DataSource and Asset parameters to do I/O (the adapter's job).
+    """
+
+    mimetype: str = ""
+
+    def __init__(self, sres: StreamResource, desc: EventDescriptor):
+        assert (
+            self.mimetype == sres["mimetype"]
```

Comment:
It should be possible for one handler to support more than one mimetype. (Perhaps it should be a `set`?) This could come up if we rename something, or if there are a range of acceptable conventions, such as `application/x-hdf` and `application/x-hdf5`.

Reply:
Sure, changed the definition to `mimetype: Union[str, Set[str]]`.
What about 'DataSource' in Tiled? Should it also accept sets of mimetypes or simply strings?

Then, it also seems to me that there should be _two_ mimetype attributes associated with each StreamHandler: one for handler discovery (possibly a `set`), and another -- the actual mimetype determined by the StreamResource (always a string, matching/included in the first mimetype attribute) and passed to DataSource. I will think how to simplify this.

Reply:
I allow a (sub-)class of `StreamHandlerBase` to accept several mimetypes (from a set) but overwrite the same field `self.mimetype` with a specific string value passed as mimetype in the Stream Resource. I'm not completely happy about the overwrite, but in this case we still have a single attribute responsible for mimetype, it contains only the relevant value for a specific instance, and -- before instantiation -- allows for handler discovery based on sets of mimetypes.